param.go

```diff
@@ -0,0 +1,189 @@
+//                param.
+type (
+	param interface {
+		// Comprehensive list of dependencies this parameter represents.
+		Dependencies() []edge
+	}
+
+	// paramList holds all arguments of the constructor as params.
+	paramList struct {
+		ctype reflect.Type // type of the constructor
+
+		Params []param
+	}
+
+	// paramSimple is an explicitly requested type, optionally with a name.
+	//
+	// This object must be present in the graph as-is unless it's specified as
+	// optional.
+	paramSimple struct {
+		Name     string
+		Optional bool
+		Type     reflect.Type
+	}
+
+	// paramObject is a dig.In struct where each field is another param.
+	//
+	// This object is not expected in the graph as-is.
+	paramObject struct {
+		Type   reflect.Type
+		Fields []paramObjectField
+
+		deps []edge
+	}
+
+	// paramObjectField is a single field of a dig.In struct.
+	paramObjectField struct {
+		// Name of the field in the struct.
+		//
+		// To clarify, this is the name of the *struct field*, not the name of
+		// the dig value requested by this field.
+		Name string
+
+		// Index of this field in the target struct.
+		//
+		// We need to track this separately because not all fields of the
+		// struct map to params.
+		Index int
+
+		// The dependency requested by this field.
+		Param param
+	}
+)
+
+var (
+	_ param = paramSimple{}
+	_ param = paramObject{}
+	_ param = paramList{}
+)
+
+// newParamList builds a paramList from the provided constructor type.
+//
+// Variadic arguments of a constructor are ignored and not included as
+// dependencies.
+func newParamList(ctype reflect.Type) (paramList, error) {
+	numArgs := ctype.NumIn()
+	if ctype.IsVariadic() {
+		// NOTE: If the function is variadic, we skip the last argument
+		// because we're not filling variadic arguments yet. See #120.
+		numArgs--
+	}
+
+	pl := paramList{
+		ctype:  ctype,
+		Params: make([]param, 0, numArgs),
+	}
+
+	for i := 0; i < numArgs; i++ {
+		p, err := newParam(ctype.In(i))
+		if err != nil {
+			return pl, errWrapf(err, "bad argument %d", i+1)
+		}
+		pl.Params = append(pl.Params, p)
+	}
+	return pl, nil
+}
+
+// newParam builds a param from the given type. If the provided type is a
+// dig.In struct, an paramObject will be returned.
+func newParam(t reflect.Type) (param, error) {
+	if IsIn(t) {
+		return newParamObject(t)
+	}
+	return paramSimple{Type: t}, nil
+}
+
+// newParamObject builds an paramObject from the provided type. The type MUST
+// be a dig.In struct.
+func newParamObject(t reflect.Type) (paramObject, error) {
+	op := paramObject{Type: t}
+
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		if f.Type == _inType {
+			// Skip over the dig.In embed.
+			continue
+		}
+
+		if f.PkgPath != "" {
+			return op, fmt.Errorf(
+				"private fields not allowed in dig.In, did you mean to export %q (%v) from %v?",
+				f.Name, f.Type, t)
+		}
+
+		p, err := newParam(f.Type)
+		if err != nil {
+			return op, errWrapf(err, "bad field %q of %v", f.Name, t)
+		}
```

Comment:
Realize that we don't want to pull in `multierr` here, but it'd be really nice to validate all the fields before returning - if there are multiple errors in a single params struct, it's pretty annoying to have them flagged one at a time. Can we just string-concat an error for the whole struct?

Reply:
I agree. There are other places where the errors should be collected too. I would prefer to do that in a follow-up diff if you don't mind. That's a lot of new behavior for this diff.